src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeOperator.scala

```diff
@@ -143,90 +141,79 @@ class TezosNodeOperator(node: TezosRPCInterface) extends LazyLogging {
-        })
+  def getBlocks(network: String, minLevel: Int, maxLevel: Int, hash: String, followFork: Boolean): Try[List[Block]] = {
+    val blocksInRange = getBlock(network, hash, 0)
+      .flatMap(block => processBlocks(network, block.metadata.hash, minLevel, maxLevel))
+    // If followFork is true, create logic for that
+    val blocksFromFork = followFork match {
+      case false => Try{List[Block]()}
+      case true => Try{List[Block]()}
+    }
+    //Need to find good implementation of liftM2
+    (blocksInRange, blocksFromFork) match {
+      case (Success(rangeBlocks), Success(forkBlocks)) => Success(forkBlocks ++ rangeBlocks)
+      case (Failure(e), _)   => throw e
+      case (_ , Failure(e))  => throw e
     }
+  }
 
   /**
     * Traverses Tezos blockchain as parameterized.
     * @param network    Which Tezos network to go against
     * @param hash       Current block to fetch
     * @param minLevel   Minimum level at which to stop
     * @param maxLevel   Level at which to start collecting blocks
-    * @param blockSoFar Blocks collected so far
-    * @param followFork If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
     * @return           All collected blocks
     */
   private def processBlocks(
                              network: String,
                              hash: String,
                              minLevel: Int,
-                             maxLevel: Int,
-                             blockSoFar: List[Block] = List[Block](),
-                             followFork: Boolean
-                           ): List[Block] =
-    getBlock(network, hash) match {
-      case Success(block) =>
-        logger.info(s"Current block height: ${block.metadata.header.level}")
-        if(block.metadata.header.level == 0 && minLevel <= 0)
-          block :: blockSoFar
-        else if(block.metadata.header.level == minLevel && !followFork)
-          block :: blockSoFar
-        else if(block.metadata.header.level <= minLevel && followFork)
-          ApiOperations.fetchBlock(block.metadata.header.predecessor) match {
-            case Success(_)     => block :: blockSoFar
-            case Failure(_)     => processBlocks(network, block.metadata.header.predecessor, minLevel, maxLevel, block :: blockSoFar, followFork)
-          }
-        else if(block.metadata.header.level > maxLevel)
-          processBlocks(network, block.metadata.header.predecessor, minLevel, maxLevel, blockSoFar, followFork)
-        else if(block.metadata.header.level > minLevel)
-          processBlocks(network, block.metadata.header.predecessor, minLevel, maxLevel, block :: blockSoFar, followFork)
-        else
-          List[Block]()
-      case Failure(e) => throw e
-    }
+                             maxLevel: Int
+                           ): Try[List[Block]] = {
+    val maxOffset: Int = maxLevel - minLevel
+    val offsets = (0 to maxOffset).toList.par
```

Comment:
trying to wrap my head around this offset, isn't this supposed to be `(minLevel to maxOffset).toList.par`?

Reply:
Nope! the way it works is getBlock has an offset from the head (latest) block in the tezos node, and maxLevel (currently) is always the latest block when this function is called. So the order of blocks being fetched from the Tezos RPC call is maxLevel, maxLevel - 1, maxLevel -2...maxLevel - (maxLevel - minLevel) = minLevel. 

Essentially, what this function does is it collects blocks from the tezos node using their RPC calls from range minLevel to maxLevel (head), which is a list of Try[Block], hence the need to sequence, in case any block fetching fails

Reply:
after additional analysis, we can possibly remove the issue of having blocking operations if we guarantee that the http calls behind the `getBlock` call are handled using asynchronous clients like it's done now with akka-http.
We can take advantage of this simply by converting the `runGetQuery` and `runPostQuery` in the "node" implementation to return an `Future` result, just like we'll do with the database calls

Reply:
Okay, so if we change runGetQuery and runPostQuery to Future results (or any sufficient asynchronous type), we'd get over that concern in sequencing the parallel collection? Otherwise, the code would look safe?

Reply:
Don't forget to make the same change for fetching accounts!

Reply:
About accounts fetching, the current branch I'm working on [CONSEIL-88-async-db-calls-2](https://github.com/Cryptonomic/Conseil/tree/CONSEIL-88-async-db-calls-2) has a significant impact already on that front